---
title: "Integration of scATAC-seq with raw data as input"
author: "Zhijian Li"
date: "July 31, 2019"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(Seurat)
library(stringr)
library(magrittr)
library(readr)
library(Matrix)
library(tidyr)
library(dplyr)
library(plotly)
library(Signac)
library(cluster)
library(mclust)
library(cowplot)
library(gridExtra)
library(ArchR)
library(foreach)
library(WriteXLS)
library(caret)

source("HiddenUtils.R")
```

```{r set_parameters, echo=FALSE}
## set parameters
set.seed(42)
addArchRThreads(threads = 100)
addArchRGenome("mm10")
```

```{r load_data}
proj <- loadArchRProject(path = "../data/Fibroblast")
```


## loadd p2g links
```{r load_p2g}
df_p2g_raw <- read.csv("../data/Fibroblast/p2g_raw.csv", row.names = 1)
df_p2g_scopen <- read.csv("../data/Fibroblast/p2g_scOpen.csv", row.names = 1)
df_p2g_magic <- read.csv("../data/Fibroblast/p2g_MAGIC.csv", row.names = 1)
df_p2g_cisTopic <- read.csv("../data/Fibroblast/p2g_cisTopic.csv", row.names = 1)
df_p2g_SCALE <- read.csv("../data/Fibroblast/p2g_SCALE.csv", row.names = 1)

df_p2g_raw <- subset(df_p2g_raw, Distance > 0)
df_p2g_scopen <- subset(df_p2g_scopen, Distance > 0)
df_p2g_magic <- subset(df_p2g_magic, Distance > 0)
df_p2g_cisTopic <- subset(df_p2g_cisTopic, Distance > 0)
df_p2g_SCALE <- subset(df_p2g_SCALE, Distance > 0)

df_p2g_raw$data <- "Raw"
df_p2g_scopen$data <- "scOpen"
df_p2g_magic$data <- "MAGIC"
df_p2g_cisTopic$data <- "cisTopic"
df_p2g_SCALE$data <- "SCALE"

df <- rbind(df_p2g_raw, df_p2g_scopen, df_p2g_magic,
            df_p2g_cisTopic, df_p2g_SCALE)

p1 <- ggplot(data = df, aes(x = reorder(data, -abs(Correlation), FUN = mean), 
                            y = Correlation, fill = data)) +
  geom_violin() +
  theme_cowplot()

p1

df.plot <- df %>%
  group_by(data) %>%
  summarise(AvgAbsCor = mean(abs(Correlation)))

p2 <- ggplot(data = df.plot, aes(x = reorder(data, -AvgAbsCor), 
                                y = AvgAbsCor, fill = data)) +
  geom_bar(stat = "identity") +
  theme_cowplot()

p2
```


## get Runx1 binding sites
```{r}
# motifInPeaks <- readRDS("../data/Fibroblast/Annotations/Motif-Matches-In-Peaks.rds")
# Runx1InPeaks <- motifInPeaks[, grep("Runx1", colnames(motifInPeaks))]
# peaksWithRunx1 <- which(as.data.frame(Runx1InPeaks@assays@data$matches)$Runx1_712)
Runx1BindingSites <- read.table("../../HINT/FootprintingFibroblast/RUNX1.bed")
gr <- GRanges(seqnames = Runx1BindingSites$V1,
              IRanges(start = as.numeric(Runx1BindingSites$V2),
                      end = as.numeric(Runx1BindingSites$V3)))
```


## load DE genes
```{r}
library(DESeq2)

load("../../../RUNX1_RNA/results/star_salmon/deseq2_qc/deseq2.dds.RData")
de_res <- results(dds) %>%
  as.data.frame() %>%
    subset(., !is.na(padj))

#de_res$DE <- ifelse(de_res$padj < 1e-05 & abs(de_res$log2FoldChange) > 1.5, 1, 0)
de_res$DE <- ifelse(de_res$padj < 1e-05 & abs(de_res$log2FoldChange) > 1, 1, 0)
de_res$gene <- rownames(de_res)
```


## define function
```{r}
library(PRROC)

get_evaluate <- function(p2g_list, final_prediction){
    pr_list <- lapply(seq_along(p2g_list), function(x){
    df_p2g <- p2g_list[[x]]
    df_p2g$gene <- toupper(df_p2g$gene)
    
    method <- names(p2g_list)[[x]]
    
    df_p2g$peak_chr <- stringr::str_split_fixed(df_p2g$peak, "_", 3)[, 1]
    df_p2g$peak_start <- stringr::str_split_fixed(df_p2g$peak, "_", 3)[, 2]
    df_p2g$peak_end <- stringr::str_split_fixed(df_p2g$peak, "_", 3)[, 3]
    peaks <- GRanges(seqnames = df_p2g$peak_chr,
                         IRanges(start = as.numeric(df_p2g$peak_start),
                                 end = as.numeric(df_p2g$peak_end)))
    o <- findOverlaps(query = peaks, subject = gr,
                          ignore.strand = TRUE)
    runx1_p2g <- df_p2g[as.integer(o@from), ]
    
    runx1_p2g <- runx1_p2g %>%
        group_by(gene) %>%
        summarise(Correlation = abs(max(Correlation))) %>%
        as.data.frame()
    
    rownames(runx1_p2g) <- runx1_p2g$gene
    commonGene <- intersect(runx1_p2g$gene, de_res$gene)
    
    de_res2 <- de_res[commonGene, ]
    runx1_p2g <- runx1_p2g[commonGene, ]

    df <- merge.data.frame(de_res2, runx1_p2g)
    
    fg <- df[df$DE == 1, ]
    bg <- df[df$DE == 0, ]

    pr <- pr.curve(scores.class0 = fg$Correlation, 
                   scores.class1 = bg$Correlation, curve = T)
    
    pr$backgroud <- nrow(fg) / (nrow(fg) + nrow(bg))
    
    pr
})
    
    df <- lapply(seq_along(pr_list), function(x){
        pr <- pr_list[[x]]
        method <- names(p2g_list)[[x]]
      
          curve <- as.data.frame(pr$curve[, 1:2])
          colnames(curve) <- c("Recall", "Precision")
          curve$Method <- method
        
        curve
    })%>% Reduce(rbind, .)

    df.aupr <- lapply(seq_along(pr_list), function(x){
      pr <- pr_list[[x]]
        method <- names(p2g_list)[[x]]
        
        aupr <- pr$auc.integral
        
        df <- data.frame(aupr = aupr, method = method)
    })%>% Reduce(rbind, .)

    background_pre <- pr_list[[1]]$backgroud
    
    return(list(df, df.aupr, background_pre))
}
```

## evaluation of peak-to-gene links using DE genes
```{r, fig.height=5, fig.width=10}

res_list <- get_evaluate(p2g_list, final_prediction = "abs_max")

df <- res_list[[1]]
df.aupr2 <- res_list[[2]]
background_pre <- res_list[[3]]

p1 <- ggplot(data = df, aes(x = Recall, y = Precision)) +
  geom_line(aes(color = Method)) +
  geom_hline(yintercept = background_pre) +
  theme_cowplot()+
    ggtitle("abs(max)")


p2 <- ggplot(data = df.aupr2, aes(x =reorder(method, -aupr), 
                                 y = aupr, fill = method)) +
  geom_bar(stat = "identity")+
    theme_cowplot() +
        xlab("") +
    theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
    ggtitle("abs(max)")


p1 + p2

saveRDS(res_list, file = "../data/Fibroblast/abs_max.Rds")
```

## Session information
```{r}
sessionInfo()
```
