---
title: "Integration of scATAC-seq with raw data as input"
author: "Zhijian Li"
date: "July 31, 2019"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(Seurat)
library(stringr)
library(magrittr)
library(readr)
library(Matrix)
library(tidyr)
library(dplyr)
library(plotly)
library(Signac)
library(cluster)
library(mclust)
library(cowplot)
library(gridExtra)
library(ArchR)
library(foreach)
library(WriteXLS)
library(caret)
```

```{r set_parameters, echo=FALSE}
## set parameters
set.seed(42)
addArchRThreads(threads = 100)
addArchRGenome("mm10")
```


```{r load_data}
proj <- loadArchRProject(path = "../data/Fibroblast")

geneMatrix <- getMatrixFromProject(proj, useMatrix = "GeneScoreMatrix")
peakMatrix <- getMatrixFromProject(proj, useMatrix = "PeakMatrix")
```


## loadd p2g links
```{r load_p2g}
df_p2g <- read.csv("../data/Fibroblast/p2g_scOpen.csv", row.names = 1)
df_p2g <- subset(df_p2g, Distance > 0 & FDR < 0.05 & Correlation > 0)

df_p2g$peak_chr <- stringr::str_split_fixed(df_p2g$peak, "_", 3)[, 1]
df_p2g$peak_start <- stringr::str_split_fixed(df_p2g$peak, "_", 3)[, 2]
df_p2g$peak_end <- stringr::str_split_fixed(df_p2g$peak, "_", 3)[, 3]
gr_peaks <- GRanges(seqnames = df_p2g$peak_chr,
                         IRanges(start = as.numeric(df_p2g$peak_start),
                                 end = as.numeric(df_p2g$peak_end)))
```


## get Runx1 binding sites
```{r}
Runx1BindingSites <- read.table("../../HINT/FootprintingFibroblast/RUNX1.bed")
gr_runx1 <- GRanges(seqnames = Runx1BindingSites$V1,
              IRanges(start = as.numeric(Runx1BindingSites$V2),
                      end = as.numeric(Runx1BindingSites$V3)))
```

## find overlapping
```{r}
o <- findOverlaps(query = gr_peaks, subject = gr_runx1,
                  ignore.strand = TRUE)

## it's possible that one peak can have multiple runx1 binding sites
df_p2g <- df_p2g[unique(as.integer(o@from)), ]
df_p2g$gene <- toupper(df_p2g$gene)

rownames(df_p2g) <- paste0(df_p2g$peak, "_", df_p2g$gene)
```

## load p2g links generated by using raw data
```{r}
df_p2g_raw <- read.csv("../data/Fibroblast/p2g_raw.csv", row.names = 1)
df_p2g_raw$gene <- toupper(df_p2g_raw$gene)
rownames(df_p2g_raw) <- paste0(df_p2g_raw$peak, "_", df_p2g_raw$gene)

df_p2g_raw <- df_p2g_raw[rownames(df_p2g), ]
```


## load data
```{r}
peaks <- as.data.frame(peakMatrix@rowRanges)
geneSet <- geneMatrix@elementMetadata

groupMatATAC_raw <- readRDS(file = "../data/Fibroblast/aggr_raw.rds")
rownames(groupMatATAC_raw) <- paste0(peaks$seqnames, "_", 
                                 peaks$start, "_", 
                                 peaks$end)

groupMatATAC_scOpen <- readRDS(file = "../data/Fibroblast/aggr_scOpen.rds")
rownames(groupMatATAC_scOpen) <- paste0(peaks$seqnames, "_", 
                                 peaks$start, "_", 
                                 peaks$end)

groupMatRNA <- readRDS(file = "../data/Fibroblast/aggr_rna.rds")
rownames(groupMatRNA) <- toupper(geneSet$name)

stopifnot(nrow(groupMatATAC_scOpen) == nrow(groupMatATAC_raw))
```


## plot individual link
```{r}
if(!dir.exists("../data/Fibroblast/Raw")){
    dir.create("../data/Fibroblast/Raw")
}

if(!dir.exists("../data/Fibroblast/scOpen")){
    dir.create("../data/Fibroblast/scOpen")
}

for (i in 1:nrow(df_p2g)) {
    gene <- df_p2g$gene[[i]]
    peak <- df_p2g$peak[[i]]
    
    correlation1 <- df_p2g$Correlation[[i]]
    correlation2 <- df_p2g_raw$Correlation[[i]]
    
    rna <- scale(groupMatRNA[gene, ])
    atac_raw <- scale(groupMatATAC_raw[peak, ])
    atac_scopen <- scale(groupMatATAC_scOpen[peak, ])

    df <- data.frame(rna = rna,
                     atac_raw = atac_raw,
                     atac_scopen = atac_scopen)
    
    df$PseudoTime <- 1:nrow(df)

    pal <- rev(ArchR::paletteContinuous(set = "blueYellow"))

    p1 <- ggplot(data = df, aes(x = rna, y = atac_scopen,
                               color = PseudoTime)) +
        geom_point() +
        scale_color_gradientn(colours = pal) +
        xlab("RNA") + ylab("ATAC") +
        ggtitle(glue::glue("Cor: {correlation1}")) +
        theme_cowplot()

    p2 <- ggplot(data = df, aes(x = rna, y = atac_raw,
                               color = PseudoTime)) +
        geom_point() +
        scale_color_gradientn(colours = pal) +
        xlab("RNA") + ylab("ATAC") +
        ggtitle(glue::glue("Cor: {correlation2}")) +
        theme_cowplot()
        
    write.table(df, file = glue::glue("../data/Fibroblast/scOpen/{gene}_{peak}.txt"),
                quote = FALSE, sep = "\t", row.names = FALSE)
    
    pdf(file = glue::glue("../data/Fibroblast/scOpen/{gene}_{peak}.pdf"),
        width = 5, height = 4)
    print(p1)
    dev.off()
    
    pdf(file = glue::glue("../data/Fibroblast/Raw/{gene}_{peak}.pdf"),
        width = 5, height = 4)
    print(p2)
    dev.off()
}
```



## Session information
```{r}
sessionInfo()
```
