---
title: "Integration of scATAC-seq with raw data as input"
author: "Zhijian Li"
date: "July 31, 2019"
output: html_document
---

```{r setup, include=FALSE}
library(ggplot2)
library(stringr)
library(magrittr)
library(cowplot)
library(ArchR)
library(Gviz)
library(cicero)
library(rtracklayer)
library(stringr)
library(GenomicRanges)
library(genomation)
```

## define function
```{r}
    gene_model <- readGFF("../../../Reference/refdata-cellranger-atac-mm10-1.2.0/genes/genes.gtf")
    gene_model$chromosome <- gene_model$seqid
    gene_model$gene <- gene_model$gene_id
    gene_model$transcript <- gene_model$transcript_id
    gene_model$symbol <- gene_model$gene_name
    gene_model$exon <- gene_model$exon_id
    gene_model$width <- gene_model$end - gene_model$start + 1
    gene_model$feature <- gene_model$transcript_type
    gene_model <- subset(gene_model, !is.na(transcript) & !is.na(exon))
    


```

## load bigwig files
```{r}
data_track_ylim <- c(0, 4)

dt_t1 <- DataTrack(range = "../../PseudoBulkTrajectory/BigWig/T1.bw", 
                   genome = "mm10", type = "h", name = "T1", col = "#F8FA0D",
                   ylim = data_track_ylim)

dt_t2 <- DataTrack(range = "../../PseudoBulkTrajectory/BigWig/T2.bw", 
                   genome = "mm10", type = "h", name = "T2", col = "#F8BA43",
                   ylim = data_track_ylim)

dt_t3 <- DataTrack(range = "../../PseudoBulkTrajectory/BigWig/T3.bw", 
                   genome = "mm10", type = "h", name = "T3", col = "#2DB7A3",
                   ylim = data_track_ylim)

dt_t4 <- DataTrack(range = "../../PseudoBulkTrajectory/BigWig/T4.bw", 
                   genome = "mm10", type = "h", name = "T4", col = "#0262E0",
                   ylim = data_track_ylim)

dt_t5 <- DataTrack(range = "../../PseudoBulkTrajectory/BigWig/T5.bw", 
                   genome = "mm10", type = "h", name = "T5", col = "#352A86",
                   ylim = data_track_ylim)

df_tf <- read.table("../../HINT/FootprintingFibroblast/RUNX1.bed",
                    header = FALSE)

colnames(df_tf) <- c("chromosome", "start", "end", "name", "score", "strand")

gr_tf <- GRanges(seqnames = df_tf$chromosome,
                     ranges = IRanges(start = df_tf$start,
                                      end = df_tf$end))
```

## load gene
```{r}
## load gene information
proj <- loadArchRProject("../data/Fibroblast")
geneMatrix <- getMatrixFromProject(proj, useMatrix = "GeneScoreMatrix")
df_gene <- as.data.frame(rowData(geneMatrix))

df_gene$gene_pos <- paste(df_gene$seqnames, 
                        df_gene$start, 
                        df_gene$start + 1,
                        sep = "_")

df_gene <- subset(df_gene, select = c("name", "gene_pos")) 
colnames(df_gene) <- c("gene", "gene_pos")
```


## define function
```{r}
plot_links <- function(df_links_p2g,
                       coaccess_cutoff_p2g,
                       df_links_p2p_1,
                       coaccess_cutoff_p2p_1,
                       df_links_p2p_2,
                       coaccess_cutoff_p2p_2,
                       sel_gene) {
    df_links_p2g <- subset(df_links_p2g, gene == sel_gene)
    
    peak_chr <-
        stringr::str_split_fixed(df_links_p2g$peak, "_", 3)[, 1]
    peak_start <-
        stringr::str_split_fixed(df_links_p2g$peak, "_", 3)[, 2]
    peak_end <-
        stringr::str_split_fixed(df_links_p2g$peak, "_", 3)[, 3]
    
    ## highlight binding sites
    gr_peaks <- GRanges(seqnames = peak_chr,
                        ranges = IRanges(start = as.numeric(peak_start),
                                         end = as.numeric(peak_end)))
    gr_overlap <- findOverlaps(
        query = gr_tf,
        subject = gr_peaks,
        type = "any",
        ignore.strand = TRUE
    )
    df_tf_sub <- df_tf[gr_overlap@from,]
    gr_tf_sub <- gr_tf[gr_overlap@from,]
    df_tf_sub$symbol <- ""
    runx1_track <- GeneRegionTrack(
        range = df_tf_sub,
        genome = "mm10",
        name = "Runx1",
        shape = "box",
        col = "blue",
        fill = "blue",
        fontsize = 12,
        fontcolor = "black"
    )
    
    # genomic coordinates
    chr <- unique(peak_chr)
    minbp <- min(as.numeric(df_tf_sub$start)) - 10000
    maxbp <- max(as.numeric(df_tf_sub$end)) + 10000
    
    
    # subset peak-to-peak links
    gene_pos <- unique(df_links_p2g$gene_pos)
    
    o <- find_overlapping_coordinates(c(df_links_p2p_1$Peak1,
                                        df_links_p2p_1$Peak2),
                                      gene_pos,
                                      maxgap = 1000)
    df_links_p2p_1 <- df_links_p2p_1[df_links_p2p_1$Peak1 %in% o |
                                         df_links_p2p_1$Peak2 %in% o,]
    df_links_p2p_1 <-
        subset(df_links_p2p_1, coaccess > coaccess_cutoff_p2p_1)
    
    o <- find_overlapping_coordinates(c(df_links_p2p_2$Peak1,
                                        df_links_p2p_2$Peak2),
                                      gene_pos,
                                      maxgap = 1000)
    df_links_p2p_2 <- df_links_p2p_2[df_links_p2p_2$Peak1 %in% o |
                                         df_links_p2p_2$Peak2 %in% o,]
    df_links_p2p_2 <-
        subset(df_links_p2p_2, coaccess > coaccess_cutoff_p2p_1)
    
    # resize peaks for peak-to-gene links
    peak_center <-
        (as.numeric(peak_start) + as.numeric(peak_end)) / 2
    df_links_p2g$peak_pos <- paste(peak_chr,
                                   peak_center,
                                   peak_center + 1,
                                   sep = "_")
    df_links_p2g <-
        subset(df_links_p2g,  select = c("peak_pos", "gene_pos", "FDR"))
    colnames(df_links_p2g) <- c("Peak1", "Peak2", "coaccess")
    df_links_p2g$coaccess <- -log10(df_links_p2g$coaccess)
    
    p2g_track <- plot_connections(
        connection_df = df_links_p2g,
        gene_model = gene_model,
        gene_model_shape = "smallArrow",
        collapseTranscripts = "longest",
        chr = chr,
        minbp = minbp,
        maxbp = maxbp,
        alpha_by_coaccess = FALSE,
        connection_ymax = max(df_links_p2g$coaccess),
        coaccess_cutoff = coaccess_cutoff_p2g,
        connection_width = 1,
        include_axis_track = FALSE,
        return_as_list = TRUE
    )
    
    p2p_track1 <- plot_connections(
        connection_df = df_links_p2p_1,
        gene_model = gene_model,
        gene_model_shape = "smallArrow",
        collapseTranscripts = "longest",
        chr = chr,
        minbp = minbp,
        maxbp = maxbp,
        alpha_by_coaccess = FALSE,
        connection_ymax = max(df_links_p2p_1$coaccess),
        coaccess_cutoff = coaccess_cutoff_p2p_1,
        connection_width = 1,
        include_axis_track = FALSE,
        return_as_list = TRUE
    )
    
     p2p_track2 <- plot_connections(
        connection_df = df_links_p2p_2,
        gene_model = gene_model,
        gene_model_shape = "smallArrow",
        collapseTranscripts = "longest",
        chr = chr,
        minbp = minbp,
        maxbp = maxbp,
        alpha_by_coaccess = FALSE,
        connection_ymax = max(df_links_p2p_1$coaccess),
        coaccess_cutoff = coaccess_cutoff_p2p_1,
        connection_width = 1,
        include_axis_track = FALSE,
        return_as_list = TRUE
    )   
    
    link_track1 <- p2g_track[[1]]
    peak_track1 <- p2g_track[[2]]
    gene_model_track <- p2g_track[[3]]
    
    link_track2 <- p2p_track1[[1]]
    peak_track2 <- p2p_track1[[2]]
    
    link_track3 <- p2p_track2[[1]]
    peak_track3 <- p2p_track2[[2]]
    
    link_track1@dp@pars$fontsize <- 12
    link_track2@dp@pars$fontsize <- 12
    link_track3@dp@pars$fontsize <- 12
    
    dt_t1@dp@pars$fontsize <- 12
    dt_t2@dp@pars$fontsize <- 12
    dt_t3@dp@pars$fontsize <- 12
    dt_t4@dp@pars$fontsize <- 12
    dt_t5@dp@pars$fontsize <- 12
    gene_model_track@dp@pars$fontsize <- 12
    
    trackList <- list(
        link_track1,
        peak_track1,
        link_track2,
        peak_track2,
        link_track3,
        peak_track3,
        dt_t1,
        dt_t2,
        dt_t3,
        dt_t4,
        dt_t5,
        gene_model_track,
        runx1_track
    )
    
    plotTracks(
        trackList,
        title.width = 0.5,
        showTitle = TRUE,
        from = minbp,
        to = maxbp,
        chromosome = chr,
        sizes = c(0.2, 0.05, 0.2, 0.05, 0.2, 0.05,
                  0.1, 0.1, 0.1, 0.1, 0.1,
                  0.1, 0.05),
        transcriptAnnotation = "symbol",
        background.title = "white",
        col.border.title = "transparent",
        lwd.border.title = "transparent",
        col.axis = "black",
        fontcolor.legend = "black",
        innerMargin = 3
    )
}


```

## load data
```{r}
df_links_p2g <- read.csv("../data/Fibroblast/p2g_scOpen.csv", row.names = 1)
df_links_p2g <- subset(df_links_p2g, Distance > 0 & FDR < 0.05 & Correlation > 0)
df_links_p2g <- merge.data.frame(df_links_p2g, df_gene)

df_links_p2p_scopen <- read.table("../../Cicero/scOpen/Cicero.txt", 
                                  header = TRUE) %>%
    subset(., coaccess > 0)

df_links_p2p_cistopic <- read.table("../../Cicero/cisTopic/Cicero.txt", 
                                  header = TRUE) %>%
    subset(., coaccess > 0)

df_links_p2p_raw <- read.table("../../Cicero/Raw/Cicero.txt", 
                                  header = TRUE) %>%
    subset(., coaccess > 0)
```



## plot for Tgfbr1
```{r, fig.height=8, fig.width=6}
plot_links(df_links_p2g = df_links_p2g, 
           coaccess_cutoff_p2g = 3,
           df_links_p2p_1 = df_links_p2p_scopen, 
           coaccess_cutoff_p2p_1 = 0.2,
           df_links_p2p_2 = df_links_p2p_raw, 
           coaccess_cutoff_p2p_2 = 0.0,
           sel_gene = "Tgfbr1")

plot_links(df_links_p2g = df_links_p2g, 
           coaccess_cutoff_p2g = 3,
           df_links_p2p_1 = df_links_p2p_scopen, 
           coaccess_cutoff_p2p_1 = 0.2,
                      df_links_p2p_2 = df_links_p2p_raw, 
           coaccess_cutoff_p2p_2 = 0.0,
           sel_gene = "Twist2")
```


## Session information
```{r}
sessionInfo()
```
